clients/include/device_batch_vector.hpp

```diff
@@ -0,0 +1,236 @@
+//
+// Copyright 2018-2019 Advanced Micro Devices, Inc.
+//
+#pragma once
+
+#include "d_vector.hpp"
+
+template <typename T>
+class host_batch_vector;
+
+//!
+//! @brief  pseudo-vector subclass which uses a batch of device memory pointers and
+//!  - an array of pointers in host memory
+//!  - an array of pointers in device memory
+//!
+template <typename T, size_t PAD = 4096, typename U = T>
+class device_batch_vector : private d_vector<T, PAD, U>
+{
+public:
+    //!
+    //! @brief Disallow copying.
+    //!
+    device_batch_vector(const device_batch_vector&) = delete;
+
+    //!
+    //! @brief Disallow or assigning.
+    //!
+    device_batch_vector& operator=(const device_batch_vector&) = delete;
+
+    //!
+    //! @brief Constructor.
+    //! @param batch_count The number of vectors.
+    //! @param size_vector The size of each vectors.
+    //!
+    explicit device_batch_vector(rocblas_int batch_count, size_t size_vector)
+        : m_batch_count(batch_count)
+        , m_size_vector(size_vector)
+        , d_vector<T, PAD, U>(size_vector)
+    {
+        if(false == this->try_initialize_memory())
+        {
+            this->free_memory();
+        }
+    };
+
+    //!
+    //! @brief Destructor.
+    //!
+    inline ~device_batch_vector() noexcept
+    {
+        this->free_memory();
+    }
+
+    //!
+    //! @brief Returns the size of the vectors.
+    //!
+    inline size_t size() const
+    {
+        return this->m_size_vector;
+    };
+
+    //!
+    //! @brief Returns the value of batch_count.
+    //!
+    inline rocblas_int batch_count() const noexcept
+    {
+        return this->m_batch_count;
+    };
+
+    //!
+    //! @brief Access to device data.
+    //! @return Pointer to the device data.
+    //!
+    inline T** ptr_on_device() noexcept
+    {
+        return this->m_device_data;
+    }
+
+    //!
+    //! @brief Const access to device data.
+    //! @return Const pointer to the device data.
+    //!
+    inline const T* const* ptr_on_device() const noexcept
+    {
+        return this->m_device_data;
+    }
+
+    //!
+    //! @brief Random access.
+    //! @param batch_index The batch index.
+    //! @return Pointer to the array on device.
+    //!
+    inline T* operator[](rocblas_int batch_index) noexcept
+    {
+        assert(0 <= batch_index && this->m_batch_count > batch_index);
+
+        return this->m_data[batch_index];
+    }
+
+    //!
+    //! @brief Constant random access.
+    //! @param batch_index The batch index.
+    //! @return Constant pointer to the array on device.
+    //!
+    inline const T* operator[](rocblas_int batch_index) const noexcept
+    {
+        assert(0 <= batch_index && this->m_batch_count > batch_index);
+
```

Comment:
`assert()` should never be used to catch user errors -- it becomes a no-op in release mode.

If it needs to still be there in release mode, some other way of indicating error is needed (but Google tests makes it hard since it doesn't use exceptions to indicate or throw errors).

Reply:
Would it be better?

return (0 <= batch_index && this->m_batch_count > batch_index)
 ? m_data[batch_index]
 :  nullptr;

Reply:
It depends on what the desired behavior is.

1. If you only want to check bounds during debug builds, use `assert()`.
2. If you want to check bounds with an error exit, use `abort()` or `exit()`, possibly with an error message.
3. If you want to check bounds using exceptions **(not detected by Google Tests)**, use `throw`.
4. (Not applicable here.) If you want to check bounds returning an error code, return a `rocblas_status_invalid_size` or other error where applicable.

Reply:
Originally I want 1.
Can we live with 1?